### PR TITLE
Center rights2

### PR DESCRIPTION
**Pull request by julien-louis**
Sorry I duplicated this pr because I wanted to do the develop merge on my computer.

linked to  https://github.com/fli-iam/shanoir-ng/issues/1148

### CODE REVIEW

**Review by Nunien — COMMENTED** [review]
Hi @jcomedouteau , I made a lot comments but some will probably turn out to be unjustified after your explanations. The task is obviously difficult and I have some improvements and questions that I want to discuss.
* As the task is hard and holes could easily appear in the security, we should extend security unit tests to the center filtering. The ideal would be to cover any method that returns affected data. I could do it if necessary.
* What is the scope of affected data ? What data can be hidden when you don't have rights on this center ?
    * datasets
    * dataset acquisitions
    * examinations
    * NOT center ? I mean for now a user can see any center (name, address, ...) the present feature doesn't change that or does it ?
* Does it affect EXPERTS as well ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java [hunk 1]
```diff
@@ -674,17 +688,46 @@ public boolean filterExaminationDTOPage(Page<ExaminationDTO> page, String rightS
...
-    	});
-    	Set<Long> checkedIds = commService.hasRightOnStudies(studyIds, rightStr);
-    	list.removeIf((ExaminationDTO exam) -> exam.getStudyId() == null || !checkedIds.contains(exam.getStudyId()));
+    	}
+    	list.removeAll(examsToRemove);
     	return true;
     }
-   
+
+    public boolean filterExaminationByCenter(List<Examination> list, Long studyId) {
```
[the remark this answers is not in the document]
Comment (jcomedouteau) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java:701]: This difference is the associated right on filterExaminationList that is not present in this method

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java [hunk 2]
```diff
@@ -217,7 +221,34 @@ public Page<Dataset> findPage(final Pageable pageable) {
 			Long userId = KeycloakUtil.getTokenUserId();
 			List<Long> studyIds = rightsRepository.findDistinctStudyIdByUserId(userId, StudyUserRight.CAN_SEE_ALL.getId());
 
-			return repository.findByDatasetAcquisitionExaminationStudyIdIn(studyIds, pageable);
+			// Check if user has restrictions.
```
[the remark this answers is not in the document]
Comment (jcomedouteau) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetServiceImpl.java:224]: as discussed, this does not work if a center is present on two studies but the user has access to this center only in one of these 2 studies.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/examination/repository/ExaminationRepository.java [hunk 3]
```diff
@@ -61,17 +62,30 @@ public interface ExaminationRepository extends PagingAndSortingRepository<Examin
 	 */
 	List<Examination> findByStudyIdIn(List<Long> studyIds);
 
+	@Query(value = "SELECT * FROM examination e WHERE e.study_id in #studyIds "
```
[the remark this answers is not in the document]
Comment (jcomedouteau) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/examination/repository/ExaminationRepository.java:65]: Yes it does with the countQuery

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/examination/repository/ExaminationRepository.java [hunk 4]
```diff
@@ -61,17 +62,30 @@ public interface ExaminationRepository extends PagingAndSortingRepository<Examin
...
 	 * @param studyIds
 	 *            list of study ids.
-	 * @param pageable
+	 * @param sort
 	 *            pagination data.
 	 * @return list of examinations.
 	 */
 	Page<Examination> findByPreclinicalAndStudyIdIn(Boolean preclinical, List<Long> studyIds, Pageable pageable);
 
+	List<Examination> findByPreclinicalAndStudyIdIn(Boolean preclinical, List<Long> studyIds, Sort sort);
```
Comment (Nunien) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/examination/repository/ExaminationRepository.java:87]: could 

`Page<Examination>findByStudyIdInFilterByCenter(List<Long> studyIds, Pageable pageable);`

replace both 

`Page<Examination> findByPreclinicalAndStudyIdIn(Boolean preclinical, List<Long> studyIds, Pageable pageable);`

and

`List<Examination> findByPreclinicalAndStudyIdIn(Boolean preclinical, List<Long> studyIds, Sort sort);`

if we add a preclinical option to it ?

Comment (jcomedouteau) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/examination/repository/ExaminationRepository.java:87]: What does "FilterByCenter" part is doing ? Maybe we should not mix Page result and List results

**Comment by jcomedouteau** [comment]
I made a new pull request on your branch with
- the updated SQL
- the merge of develop

**Comment by jcomedouteau** [comment]
I have some problems on
- Examination list -> Access denied (even if I should have access to some exams)
- Dataset list -> I can only see one dataset instead of multiple (this is really quite weird -> Maybe linked to the SQL I changed)

**Comment by julien-louis** [comment]
I think it's good now

**Comment by julien-louis** [comment]
I think it's good now !

**Comment by jcomedouteau** [comment]
Let's deploy in qualif and test ?

**Comment by julien-louis** [comment]
Quick test : 
* could add a user to a multi-center study and edit his center rights
* once logged the user can only see those centers in solr

**Pull request merged**